File: .scratch-t11/preview.py
"""Preview: t11's two band constructions beside c45 and the shipped Twozz."""
import json
import re
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SC = ROOT / '.scratch-t11'


def face_svg(cx, cy, size, smile, gap, fill):
    js = (f"import {{facePathsAt}} from '{ROOT}/src/data/mark.ts';"
          f"console.log(JSON.stringify(facePathsAt({{cx:{cx},cy:{cy},size:'{size}',"
          f"smile:'{smile}',gap:{gap}}})));")
    ds = json.loads(subprocess.run(
        ['node', '--experimental-strip-types', '--input-type=module', '--no-warnings', '-e', js],
        capture_output=True, text=True, check=True).stdout)
    return f'<g fill="{fill}">' + ''.join(f'<path d="{d}"/>' for d in ds) + '</g>'


def from_astro(slug, face):
    src = (ROOT / f'src/components/mark/logos/{slug}.astro').read_text()
    body = '\n'.join(f'<path d="{d}" fill="{f}"/>'
                     for d, f in re.findall(r'<path d="([^"]+)"\s+fill="([^"]+)"', src))
    return (f'<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" '
            f'shape-rendering="crispEdges">{body}{face}</svg>')


c45 = from_astro('c45', face_svg(16, 13, 'md', 'compact', 2, '#132638'))
shipped = (ROOT / '.briefs/twozz-shipped.svg').read_text()
shipped = re.sub(r'width="\d+" height="\d+"', '', shipped, count=1)
a = (SC / 'a-shells.svg').read_text()
b = (SC / 'b-rings.svg').read_text()

MARKS = [('c45 (Hozz)', c45), ('t11 shells', a), ('t11 rings', b), ('shipped', shipped)]


def sized(svg, px):
    return svg.replace('<svg ', f'<svg width="{px}" height="{px}" ', 1)


def row(px, bg, fg):
    cells = ''.join(
        f'<div style="display:flex;flex-direction:column;align-items:center;gap:6px">'
        f'{sized(s, px)}<span style="font:11px system-ui;color:{fg};opacity:.65">{n}</span></div>'
        for n, s in MARKS)
    return (f'<div style="display:flex;gap:34px;align-items:flex-end;padding:22px 26px;'
            f'background:{bg};border-radius:14px;margin:10px">{cells}</div>')


html = ('<body style="margin:0;background:#e9eaee;padding:10px">'
        + row(96, '#ffffff', '#111') + row(96, '#15161a', '#eee')
        + row(48, '#ffffff', '#111') + row(24, '#ffffff', '#111')
        + row(24, '#15161a', '#eee') + row(16, '#ffffff', '#111') + '</body>')
(SC / 'p.html').write_text(html)

shot = f'''
import {{ chromium }} from '/Users/brandon/Development/amazon-subscription-canceller/node_modules/playwright/index.mjs';
const b = await chromium.launch();
const p = await b.newPage({{ viewport: {{ width: 620, height: 780 }}, deviceScaleFactor: 3 }});
await p.goto('file://{SC}/p.html');
await p.screenshot({{ path: '{SC}/p.png', fullPage: true }});
await b.close();
'''
(SC / 'shot.mjs').write_text(shot)
subprocess.run(['node', SC / 'shot.mjs'], check=True)
print('shot written')

File: .scratch-t15/preview.mjs
import fs from 'fs';
import path from 'path';
import { facePathsAt } from '../src/data/mark.ts';
import { chromium } from '/Users/brandon/Development/amazon-subscription-canceller/node_modules/playwright/index.mjs';

const ROOT = process.cwd();

function svgFromAstro(file) {
  const src = fs.readFileSync(path.join(ROOT, 'src/components/mark/logos', file), 'utf8');
  const body = src.split('<MarkFrame')[1];
  const paths = [...body.matchAll(/<path d="([^"]+)"\s+fill="([^"]+)"/g)]
    .map(([, d, f]) => `<path d="${d}" fill="${f}"/>`)
    .join('');
  const g = body.match(/<g fill="([^"]+)"/);
  const opts = body.match(/facePathsAt\(\{([^}]+)\}\)/);
  let face = '';
  if (opts) {
    const o = {};
    for (const kv of opts[1].split(',')) {
      const [k, v] = kv.split(':').map((s) => s.trim());
      o[k] = v.startsWith("'") ? v.slice(1, -1) : Number(v);
    }
    face = `<g fill="${g[1]}">` + facePathsAt(o).map((d) => `<path d="${d}"/>`).join('') + '</g>';
  }
  return (size) =>
    `<svg width="${size}" height="${size}" viewBox="0 0 32 32" shape-rendering="crispEdges" style="display:block">${paths}${face}</svg>`;
}

const shipped = fs.readFileSync(path.join(ROOT, '.briefs/twozz-shipped.svg'), 'utf8');
const shippedAt = (size) =>
  shipped.replace('width="32" height="32"', `width="${size}" height="${size}"`)
         .replace('<svg ', '<svg style="display:block" ')
         .replace(/id="c"/g, 'id="cc"').replace(/url\(#c\)/g, 'url(#cc)');

const t15 = svgFromAstro('t15.astro');
const c45 = svgFromAstro('c45.astro');

const cell = (label, svg) =>
  `<div style="display:flex;flex-direction:column;align-items:center;gap:6px">
     ${svg}<span style="font:10px system-ui;opacity:.55">${label}</span></div>`;

const row = (size, bg, fg) =>
  `<div style="display:flex;gap:26px;align-items:flex-end;padding:18px 22px;background:${bg};color:${fg};border-radius:12px;margin:10px">
     ${cell('t15 ' + size, t15(size))}
     ${cell('shipped', shippedAt(size))}
     ${cell('c45 Hozz', c45(size))}
   </div>`;

fs.writeFileSync('.scratch-t15/p.html',
  `<body style="margin:0;background:#eef0f3;font-family:system-ui">
     ${row(96, '#fbfbfc', '#111')}
     ${row(96, '#17181c', '#eee')}
     ${row(48, '#fbfbfc', '#111')}
     ${row(24, '#fbfbfc', '#111')}
     ${row(24, '#17181c', '#eee')}
     ${row(16, '#fbfbfc', '#111')}
   </body>`);

const b = await chromium.launch();
const p = await b.newPage({ viewport: { width: 520, height: 760 }, deviceScaleFactor: 3 });
await p.goto('file://' + ROOT + '/.scratch-t15/p.html');
await p.screenshot({ path: '.scratch-t15/p.png', fullPage: true });
await b.close();
console.log('shot');
